### PR TITLE
feat(edgex-secrets-config): Replace security-proxy-setup utility for adding proxy users

### PR DESCRIPTION
**Pull request by bnevis-i**
This utility replaces security-proxy-setup for adding users to the Kong proxy.
    
    This new utility changes the JWT authentication mode from HS256 to RS256,
    thus avoiding exposure of the HMAC secret on the server side.
    This new utility also puts the client_id and client_secret for
    OAuth2 authentication in the hands of the user.
    Although in the new scheme the user is responsible for generating
    tokens via either the client_credentials OAuth2 flow,
    or client-side JWT signing, utility methods are provided
    for that function that do not depend upon the server.
    
    The new utility eliminates the requirement from security-proxy-setup
    that a valid security secret store token is available.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] The commit subject follows the [Conventional Commits spec](https://github.com/zeke/semantic-pull-requests)
- [ ] The commit message follows the [EdgeX Contributor Guide](https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] `make test` has completed successfully

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number:

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by bnevis-i** [comment]
@jim-wang-intel   Jim, wondering if you have time to do a first-pass code review on this and tear into it.  I don't have any tests written yet.

**Comment by jim-wang-intel** [comment]
> 
> 
> @jim-wang-intel Jim, wondering if you have time to do a first-pass code review on this and tear into it. I don't have any tests written yet.

ok, I am reviewing it.

**Comment by bnevis-i** [comment]
@jim-wang-intel You should be able to start up the framework using docker-compose and compile this utility standalone and use it.  Note that to change from jwt to oauth authentication, you need to edit the configuration.toml for security-proxy-setup, rebuild the container, and set up edgex fresh again.

File: internal/security/config/command/proxy/jwt/command.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+	token := jwt.NewWithClaims(signingMethod, claims)
+
+	signedToken, err := token.SignedString(key)
+	if err != nil {
+		return contract.StatusCodeExitWithError, fmt.Errorf("Could not sign JWT: %w", err)
+	}
+
+	if len(c.outputPath) > 0 {
+		ioutil.WriteFile(c.outputPath, []byte(signedToken), os.FileMode(0600))
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/jwt/command.go:127]: add error handling if writing file failed

File: internal/security/config/command/proxy/oauth2/command.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	var parsedResponse map[string]interface{}
+
+	if err := json.NewDecoder(bytes.NewReader(responseBody)).Decode(&parsedResponse); err != nil {
+		return contract.StatusCodeExitWithError, fmt.Errorf("Unable to parse create token response: %w", err)
+	}
+
+	if len(c.outputPath) > 0 {
+		ioutil.WriteFile(c.outputPath, []byte(responseBody), os.FileMode(0600))
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/oauth2/command.go:112]: add error handling if writing file failed

File: internal/security/config/command/proxy/adduser/command.go [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+	case http.StatusConflict:
+		return contract.StatusCodeExitWithError, fmt.Errorf("Associate JWT request failed (likely due to duplicate ID) with code: %d", resp.StatusCode)
+	default:
+		return contract.StatusCodeExitWithError, fmt.Errorf("Associate JWT request failed with code: %d", resp.StatusCode)
+	}
+
+	outputKey := fmt.Sprintf("%s", parsedResponse["key"])
+
+	if len(c.outputIDPath) > 0 {
+		ioutil.WriteFile(c.outputIDPath, []byte(outputKey), os.FileMode(0600))
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/adduser/command.go:268]: add error handling if writing file failed

File: internal/security/config/command/proxy/adduser/command.go [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+	if err != nil {
+		return fmt.Errorf("Failed to prepare new consumer request %s: %w", c.username, err)
+	}
+	req.Header.Add(clients.ContentType, "application/x-www-form-urlencoded")
+	resp, err := c.client.Do(req)
+	if err != nil {
+		_ = resp.Body.Close()
+		return fmt.Errorf("Failed to send new consumer request %s: %w", c.username, err)
+	}
+	responseBody, err := ioutil.ReadAll(resp.Body)
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/adduser/command.go:148]: either log `err` or return `err`

File: internal/security/config/command/proxy/adduser/command.go [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+	if err != nil {
+		return fmt.Errorf("Failed to build request to associate consumer %s to group %s: %w", c.username, c.group, err)
+	}
+	req.Header.Add(clients.ContentType, "application/x-www-form-urlencoded")
+	resp, err := c.client.Do(req)
+	if err != nil {
+		_ = resp.Body.Close()
+		return fmt.Errorf("Failed to subnit request to associate consumer %s to group %s: %w", c.username, c.group, err)
+	}
+	responseBody, err := ioutil.ReadAll(resp.Body)
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/adduser/command.go:188]: either log `err` or return `err`

File: internal/security/config/command/proxy/adduser/command.go [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+		return contract.StatusCodeExitWithError, fmt.Errorf("Failed to prepare request to associate JWT to user %s: %w", c.username, err)
+	}
+	req.Header.Add(clients.ContentType, "application/x-www-form-urlencoded")
+	resp, err := c.client.Do(req)
+	defer resp.Body.Close()
+	if err != nil {
+		return contract.StatusCodeExitWithError, fmt.Errorf("Failed to send request to associate JWT to user %s: %w", c.username, err)
+	}
+
+	responseBody, err := ioutil.ReadAll(resp.Body)
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/adduser/command.go:248]: either log `err` or return `err`

File: internal/security/config/command/proxy/adduser/command.go [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+	}
+
+	return &cmd, err
+}
+
+func (c *cmd) Execute() (statusCode int, err error) {
+	switch c.tokenType {
+	case contract.JwtTokenType:
+		statusCode, err = c.ExecuteAddJwt()
+		break
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/adduser/command.go:115]: redundant, remove

File: internal/security/config/command/proxy/adduser/command.go [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+	case contract.OAuth2TokenType:
+		statusCode, err = c.ExecuteAddOAuth2()
+		break
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/adduser/command.go:118]: redundant, remove

File: internal/security/config/command/proxy/adduser/command.go [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+		return fmt.Errorf("Failed to send new consumer request %s: %w", c.username, err)
+	}
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	_ = resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+	case http.StatusCreated:
+		c.loggingClient.Info(fmt.Sprintf("created consumer (user) '%s'", c.username))
+		break
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/adduser/command.go:155]: redundant, remove

File: internal/security/config/command/proxy/adduser/command.go [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+	case http.StatusConflict:
+		c.loggingClient.Info(fmt.Sprintf("consumer '%s' already created", c.username))
+		break
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/adduser/command.go:158]: redundant, remove

**Comment by bnevis-i** [comment]
@jim-wang-intel Thank you.

**Pull request closed without merging**